### PR TITLE
Added a new cookbook about file uploading

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes |
| Applies to | all |
| Fixed tickets | #2346 |

This PR is completely based on the work made by @saro0h in #4018. All the credit for the original work goes to her. I just picked her work and applied the suggestions made by @weaverryan @xabbuh and @WouterJ to make it mergeable.

If you agree with this PR, please merge it as soon as possible because this is a very important topic and we're a bit late on this doc. Thanks!

### CODE REVIEW

File: cookbook/controller/upload_file.rst [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            $this->brochure = $brochure;
+
+            return $this;
+        }
+    }
+
+Note that the type of the ``brochure`` column is ``string`` instead of ``binary``
+or ``blob`` because it just stores the PDF file name instead of the file contents.
+
+Then, add a new ``brochure`` field to the form that manages ``Product`` entities::
```
Comment (weaverryan) [inline cookbook/controller/upload_file.rst]: ... manage the ``Product`` entity::

File: cookbook/controller/upload_file.rst [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        public function buildForm(FormBuilderInterface $builder, array $options)
+        {
+            $builder
+                // ...
+                ->add('brochure', 'file', array('label' => 'Brochure (PDF file)'))
+                // ...
+            ;
+        }
+
+        // ...
```
Comment (weaverryan) [inline cookbook/controller/upload_file.rst]: Let's include the `setDefaultOptions()` here with `data_class`, and also `getName()`. I think it'll just make the entry "complete"

File: cookbook/controller/upload_file.rst [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+                $fileName = md5(uniqid()).'.'.$file->guessExtension();
+
+                // Move the file to the directory where brochures are stored
+                $brochuresDir = $this->container->getParameter('kernel.root_dir').'/../web/uploads/brochures';
+                $file->move($brochuresDir, $fileName);
+
+                // Update the 'brochure' property to store the PDF file name instead of its contents
+                $product->setBrochure($filename);
+
+                // ...
```
Comment (weaverryan) [inline cookbook/controller/upload_file.rst]: // persist the $product variable or any other work...

File: cookbook/controller/upload_file.rst [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        /**
+         * @Route("/product/new", name="app_product_new")
+         */
+        public function newAction(Request $request)
+        {
+            //...
+
+            if ($form->isValid()) {
+                // $file stores the uploaded PDF file
+                $file = $product->getBrochure()
```
Comment (weaverryan) [inline cookbook/controller/upload_file.rst]: What do you think about adding this above?

```php
/** @var Symfony\...\UploadedFile $file */
```

**Comment by weaverryan** [comment]
:+1: from me after the changes. It's simpler than most ways this is explained, and I like that.

File: cookbook/controller/upload_file.rst [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    use AppBundle\Entity\Product;
+
+    class ProductController extends Controller
+    {
+        /**
+         * @Route("/product/new", name="app_product_new")
+         */
+        public function newAction(Request $request)
+        {
+            //...
```
Comment (weaverryan) [inline cookbook/controller/upload_file.rst]: We might as well show the `$form = $this->createForm()` part - then this entry is totally complete

**Comment by javiereguiluz** [comment]
@weaverryan thanks for your review! I've made all the suggested changes.

**Comment by weaverryan** [comment]
Thanks Javier! Now, I also think we should remove http://symfony.com/doc/current/cookbook/doctrine/file_uploads.html entirely and then here, maybe add a note /link about how you might move the upload logic to a service and maybe even use a Doctrine listener if you are in fact using Doctrine. What do you think about that? The other article is not great - this is better (even though it's shorter).

**Pull request merged**